internal/authentication/oidc/clientcredentials.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oidc
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"golang.org/x/oauth2"
+	ccreds "golang.org/x/oauth2/clientcredentials"
+
+	"github.com/dapr/kit/logger"
+)
+
+const (
+	oidcScopeOpenID = "openid"
+)
+
+type ClientCredentialsOptions struct {
```

Comment:
Can you add the `mapstructure` properties to this? So then components' metadata structs can inherit this, and we can help standardizing the metadata options should other components adopt OAuth2 authentication in the future.

Reply:
I'm actually very against this. I'm quite keen that we have a distinct and clean separation between external and internal APIs. Happy to create a shared metadata API package for this reason, but I don't think we should have external API definitions in this internal implementation package.

Reply:
The problem that we've seen is that then components implement this using different metadata keys. Before we standardized the Azure AD metadata, some components used "azureClientID", some "spnClientID", some "clientID", for the exact same property.

If you'd rather have a separate struct, that's fine, but I'd still like to have a shared one in this package.

Reply:
Done [here](https://github.com/dapr/components-contrib/pull/3026/commits/b817c7c164f621d8b208c3e2874b71f15be31a8f)